src/instrumentation.rs

```diff
@@ -0,0 +1,108 @@
+//! APIs wrapping the Python interpreter's instrumentation features.
+use crate::ffi;
+use crate::pyclass::boolean_struct::False;
+use crate::types::PyFrame;
+use crate::{Bound, PyAny, PyClass, PyObject, PyRefMut, PyResult, Python};
+use std::ffi::c_int;
+
+/// Represents a monitoring event used by the profiling API
+pub enum ProfileEvent<'py> {
+    /// A python function or method was called or a generator was entered.
+    Call,
+    /// A python function or method returned or a generator yielded. The
+    /// contained data is the value returned to the caller or `None` if
+    /// caused by an exception.
+    Return(Option<Bound<'py, PyAny>>),
+    /// A C function is about to be called. The contained data is the
+    /// function object being called.
+    CCall(Bound<'py, PyAny>),
+    /// A C function has raised an exception. The contained data is the
+    /// function object being called.
+    CException(Bound<'py, PyAny>),
+    /// A C function has returned. The contained data is the function
+    /// object being called.
+    CReturn(Bound<'py, PyAny>),
+}
+
+impl<'py> ProfileEvent<'py> {
+    fn from_raw(what: c_int, arg: Option<Bound<'py, PyAny>>) -> ProfileEvent<'py> {
+        match what {
+            ffi::PyTrace_CALL => ProfileEvent::Call,
+            ffi::PyTrace_RETURN => ProfileEvent::Return(arg),
+            ffi::PyTrace_C_CALL => ProfileEvent::CCall(arg.unwrap()),
+            ffi::PyTrace_C_EXCEPTION => ProfileEvent::CException(arg.unwrap()),
+            ffi::PyTrace_C_RETURN => ProfileEvent::CReturn(arg.unwrap()),
+            _ => unreachable!(),
+        }
+    }
+}
+
+/// Trait for Rust structs that can be used with Python's profiling API.
+pub trait Profiler: PyClass<Frozen = False> {
```

Comment:
The `Profiler` trait as a subtype of `PyClass` is an interesting choice. Is that strictly necessary? Is that for the borrow checking? I wonder how this might interact with nogil. I also whether something like what we've got with `run_closure` over in the function code might work (e.g. just register a `Fn()` closure rather than a full type).

Reply:
I think it is necessary, but maybe there's an alternative that avoids it that I missed.